### ISSUE TITLE
Util method calls pop on dead solvers

### ISSUE THREAD

**Issue by Heizmann** [ultimate-pa/ultimate#191]
The Util class has an [auxiliary checksat method](https://github.com/ultimate-pa/ultimate/blob/9b274f421a02dfa03aaa92aa2e86c75ae66f6a13/trunk/source/Library-SMTLIB/src/de/uni_freiburg/informatik/ultimate/logic/Util.java#L37-L64) that has many callers.

The method ends with the following code

		} finally {
			script.pop(1);
		}

Problem: 
* Solvers may crash with a nice error message
* The error message is perfectly encapsulated in an SMTLIBException
* After exception is thrown the _finally_ block is executed
* Since the solver is already dead the pop command fails and a new exception is thrown e.g., 
`SMTLIBException: External (z3 SMTLIB2_COMPLIANT=true -memory:2024 -smt2 -in -t:2000) Connection to SMT solver broken:`
* The core gets only the second exception and I cannot see the real reason why the solver crashed

Can we remove this try-catch-finally block?

Problem is probably reproducible with the following
Input:`svcomp/float-benchs/filter2_iterated_true-unreach-call.c`
Settings:`settings/automizer/interpolation/float/Reach-32bit-MathSAT-IcSpLv-Float-FakeNi.epf`
Toolchain:`toolchains/AutomizerCInline.xml`

**Comment by danieldietsch** [comment]
If there is no other exception it should be enough to ensure that ``pop()`` is called before each exit and if the solver is still alive. I can look into that.

**Comment by Heizmann** [comment]
There **is** another exception.
Currently, I am using a workaround for the problem that I could also commit.
Maybe we should discuss the problem via Skype/phone.

**Comment by greitsch** [comment]
How about this?

```Java

try {
   // ...
} catch (SomeException se) {
   mSolverError = true;
   //...
} finally {
   if (!mSolverError) {
      script.pop(1);
   }
}
```

Is that possible?

**Comment by Heizmann** [comment]
It solves the problem, but has no benefit over the simpler solution in which we just drop the try-catch-finally block.

**Comment by Heizmann** [comment]
I am currently using the following workaround.

	public static LBool checkSat(final Script script, Term term) {
		script.push(1);
		SMTLIBException checkSatException = null;
		try {
			final TermVariable[] vars = term.getFreeVars();
			final Term[] values = new Term[vars.length];
			for (int i = 0; i < vars.length; i++) {
				values[i] = termVariable2constant(script, vars[i]);
			}
			term = script.let(vars, values, term);
			LBool result = script.assertTerm(term);
			if (result == LBool.UNKNOWN) {
				result = script.checkSat();
			}
			return result;
		} catch (final SMTLIBException e) {
			checkSatException = e;
		} finally {
			try {
				script.pop(1);
			} catch (final SMTLIBException popException) {
				// ignore popException and throw the check-sat exception
				throw checkSatException;
			}
		}
		throw new AssertionError("I think this is unreachable, but the java compiler cannot infer this");
	}

**Comment by danieldietsch** [comment]
Perfect. Use this.

Am 15. Juni 2017 11:05:34 MESZ schrieb Matthias Heizmann <notifications@github.com>:

**Comment by danieldietsch** [comment]
This evening maybe?

Am 15.06.2017 10:40 schrieb "Matthias Heizmann" <notifications@github.com>:

**Comment by Heizmann** [comment]
> Perfect. Use this.
But why should we use a complicated workaround if we can simply remove try-catch-finally block.

**Comment by danieldietsch** [comment]
@Heizmann fixed this with b07a7f0c268614fc2fccedb2b68b5651d2ad8d32 but forgot to reference or close this ticket.
